### PR TITLE
fix(firebase_auth): Fix email link signin on Android

### LINKED ISSUE #4711: [Firebase_auth] : Email Link Sign-In doesn't work on Android
I followed [these](https://firebase.flutter.dev/docs/auth/usage/#email-link-authentication) instructions on firebase.flutter.dev to implement email-link auth in my Flutter app. The mixture of Flutter-specific instructions with links to Firebase generic instructions is quite confusing. After spending multiple days, email-link sign in is still not working. 

I successfully implemented email & password auth as well as Facebook auth using FlutterFire. My app is also successfully communicating with Firebase Firestore and Functions. I assume my project is setup correctly. I also implemented email link sign-in for my Firebase project's AngularFire web app without any issues.

Here is the method in my Flutter app that calls Firebase Auth:

```
Future<void> _signInWithEmailLink() async {
    final acs = ActionCodeSettings(
        url: "https://rebait.page.link/2irR",
        handleCodeInApp: true,
        iOSBundleId: "io.consumer.rebait.app",
        androidPackageName: "app.rebait.consumer",
        androidInstallApp: true,
        androidMinimumVersion: "12");
    try {
      await _auth.sendSignInLinkToEmail(email: _emailController.text, actionCodeSettings: acs);
      await _showEmailLinkSentDialog();
    } catch (err) {
      print('SIGN IN LINK ERROR');
      print(err.toString());
      return;
    }
    // Navigator.of(context).popUntil((route) => route.isFirst);
  }
```

Here is the function that is waiting to handle the sign-in link:

```
FirebaseDynamicLinks.instance.onLink(onSuccess: (PendingDynamicLinkData dynamicLink) async {
      print('ON LINK');
      print(dynamicLink.link);
      print(dynamicLink.link.data);
      if (FirebaseAuth.instance.isSignInWithEmailLink(dynamicLink.link.toString())) {
      if (FirebaseAuth.instance.isSignInWithEmailLink(dynamicLink.toString())) {
        print('IS SIGN IN LINK');
      } else {
        print('IS NOT SIGN IN LINK');
      }
    }, onError: (OnLinkErrorException e) async {
      print(e.message);
      _showErrorDialog(e);
    });
```

My app manages to call Firebase and an email is sent to my inbox with a sign-in link. When I click on the link in the Gmail app on my Android emulator, my app is moved to the foreground and it handles the link. The prints from `onLink` are:

```
I/flutter (22864): ON LINK
I/flutter (22864): https://rebait.page.link
I/flutter (22864): null
I/flutter (22864): IS NOT SIGN IN LINK
```

My dynamic link settings are as follows. However, I don't really understand what the purpose of this is. Doesn't Firebase Auth generate the necessary deep link data based on the ActionCodeSetting automatically (it does for AngularFire)? Not sure why I have to setup some dynamic link manually in the console.

Screen Shot 2021-01-18 at 11 41 34 AM

Firebase Auth settings:

Screen Shot 2021-01-18 at 11 43 14 AM

Screen Shot 2021-01-18 at 11 43 30 AM

Flutter doctor output:
```
[✓] Flutter (Channel stable, 1.22.5, on macOS 11.1 20C69 darwin-x64, locale en-US)
    • Flutter version 1.22.5 at /Users/arnewolframm/flutter
    • Framework revision 7891006299 (6 weeks ago), 2020-12-10 11:54:40 -0800
    • Engine revision ae90085a84
    • Dart version 2.10.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/arnewolframm/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.0.rc.1

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] Connected device (1 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86 • Android 11 (API 30) (emulator)
    ! Error: Arne’s iPhone is not connected. Xcode will continue when Arne’s iPhone is connected. (code -13)
```

Pubspec yaml:

```
name: consumer
description: Rebait Consumer App

version: 1.0.0+1

environment:
  sdk: ">=2.10.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter
  flutter_localizations:
    sdk: flutter

  cupertino_icons: ^1.0.0

  firebase_core: ^0.7.0

  firebase_analytics: ^7.0.0

  firebase_auth: ^0.20.0+1
  cloud_firestore: ^0.16.0
  cloud_functions: ^0.9.0

  google_sign_in: ^4.4.4
  provider: ^4.1.1
  firebase_dynamic_links: ^0.7.0
  flutter_signin_button: ^1.0.0
  email_validator: ^1.0.0
  package_info: '>=0.4.0 <2.0.0'
  flutter_facebook_auth: 0.2.4
  rxdart: ^0.25.0
  google_fonts: ^1.1.0
  wakelock: ^0.2.1
  json_annotation: ^3.1.1
  http: ^0.12.2
  geolocator: ^6.0.0


dev_dependencies:
  build_runner: ^1.10.1
  json_serializable: ^3.5.1

  flutter_test:
    sdk: flutter

flutter_intl:
  enabled: true
  localizely:
    project_id: [...]
```

PLEASE HELP!

### PR DESCRIPTION
**Pull request by jyardin**
## Description

Changes the `ActionCodeSettings.asMap` implementation so that it is correctly consumed in all platforms (Android/iOS/Web).
This fixes a bug where the sign in link generated on Android doesn't work. (see https://github.com/FirebaseExtended/flutterfire/issues/4711#issuecomment-773791945)

## Related Issues

Fixes #4711
It might fix other bugs related to `ActionCodeSettings`.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by jyardin** [comment]
Note on the fix: another option would be to change the way the `ActionCodeSettings` map is consumed in all platforms. But changing the `ActionCodeSettings.asMap` method seemed more inline with the intent of the original refactoring, and has less impact on the code.

**Comment by wolframm** [comment]
Do you have an ETA for this bug fix? Developers with Android phones literally cannot work with firebase_auth: ^0.20.0+1, because they cannot sign in to their apps on Android phone or emulator, and the iPhone emulator doesn't allow one to install email apps.

Only option to continue working is downgrade to firebase_auth: ^0.18.4. I need to know if I should do the work of downgrading or wait for bug fix.

**Review by russellwheatley — APPROVED** [review]
This appears to be the correct fix. Looking at our android [implementation](https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebase/auth/FlutterFirebaseAuthPlugin.java#L449-L469), it is expecting an optional nested `android`map with the "deprecated" params we removed. The android  [implementation](https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebase/auth/FlutterFirebaseAuthPlugin.java#L471-L476) also expects an optional nested `iOS` map.

The `iOS`[implementation](https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_auth/firebase_auth/ios/Classes/FLTFirebaseAuthPlugin.m#L1099-L1115) also expects an optional `iOS` &  optional`android` map.

The implementation itself is interesting, I didn't realise you could nest if conditions in a Dart map 👍 .

Tests are failing but I suspect a number of them will be mitigated once this PR: https://github.com/FirebaseExtended/flutterfire/pull/5014 lands which fixes a number of issues we've experienced as part of the null safety upgrade. So I'd suggest waiting for that to be merged before merging this PR.

**Review by Salakar — APPROVED** [review]

**Review by Salakar — APPROVED** [review]

**Pull request merged**
